pkg/action/hooks.go

```diff
@@ -0,0 +1,122 @@
+/*
+Copyright The Helm Authors.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package action
+
+import (
+	"bytes"
+	"sort"
+	"time"
+
+	"github.com/pkg/errors"
+
+	"helm.sh/helm/pkg/release"
+)
+
+// execHook executes all of the hooks for the given hook event.
+func (cfg *Configuration) execHook(rl *release.Release, hook release.HookEvent, timeout time.Duration) error {
+	executingHooks := []*release.Hook{}
+
+	for _, h := range rl.Hooks {
+		for _, e := range h.Events {
+			if e == hook {
+				executingHooks = append(executingHooks, h)
+			}
+		}
+	}
+
+	sort.Sort(hookByWeight(executingHooks))
+
+	for _, h := range executingHooks {
+		if err := cfg.deleteHookByPolicy(h, release.HookBeforeHookCreation); err != nil {
+			return err
+		}
+
+		resources, err := cfg.KubeClient.Build(bytes.NewBufferString(h.Manifest))
+		if err != nil {
+			return errors.Wrapf(err, "unable to build kubernetes object for %s hook %s", hook, h.Path)
+		}
+		if _, err := cfg.KubeClient.Create(resources); err != nil {
+			return errors.Wrapf(err, "warning: Hook %s %s failed", hook, h.Path)
+		}
+
+		// Get the time at which the hook was applied to the cluster
+		h.LastRun = release.HookExecution{
+			StartedAt: time.Now(),
+			Phase:     release.HookPhaseUnknown,
```

Comment:
can we introduce a new phase: `release.HookPhaseRunning` and use that here instead? We know it's in a running state at this time. In Kubernetes, a resource in an unknown state is used to inform the user that the resource has entered an erroneous state (i.e. Helm was unable to retrieve the hook's completion state), and the user should intervene to determine what should be done to reconcile its state.

Reply:
`HookPhaseUnknown` was used here rather than `HookPhaseRunning` because
1. No intermediate state about the release is currently persisted during execution.
1. As long as the implementation of `WatchUntilReady` does not panic `HookPhaseFailed` or `HookPhaseSucceeded` should always be set (and if there is a panic, then `HookPhaseUnknown` is the most appropriate value to surface).

0a11574b adds `HookPhaseRunning` and persists intermediate state. Is the extra API call acceptable?